### DISCUSSION TITLE
What is the recommended approach for handling transfers between accounts, such as credit card payments?

### DISCUSSION THREAD

**Discussion by zzzachzzz** [firefly-iii/firefly-iii#7313] [Firefly III] [answered]
I'll preface by saying that I'm new to Firefly, though I feel I've spent a fair amount of time reading the docs, searching GitHub issues, and searching GitHub discussions. Sorry about this maybe being considered a duplicate question, but to be honest I wasn't satisfied with the answers I was seeing in other discussions.

My question is pretty simple. What is the recommended way to handle transfers (in an automated fashion)?

Consider a couple basic scenarios, such as money being transferred from a bank checking account to a savings account, or from a checking account to a credit card for payment (which apparently is also considered an asset account, okay).

I should state that I've setup my accounts with csv imports.

For transactions that involve moving money from one asset account to another, I need to address them, so that a -$1,000 isn't treated as an expense, and likewise a +$1,000 on the other end isn't treated as income.

So I should just make it a transfer in Firefly, right? The problem I have with transfers is that when I convert a withdrawal to a transfer, the deposit transaction lingers. The advice I've seen given is to delete the deposit transaction in order to avoid that two instances of +$1,000 for this example.

Wouldn't it make more sense to have the ability for a transfer to link to both the withdrawal transaction, and the deposit transaction?

I had more success doing this with a "Is paid for by" transaction link instead of a transfer. Something doesn't feel right about deleting a transaction to compensate for the conversion of its opposite transaction to a transfer. What happens if I import an updated csv, will I need to go back and delete that transaction again, to avoid doubled amounts?

I would just use transaction links as I've described to solve this issue, but I did not see a way to do this in an automated fashion with rules. I feel weird calling it an issue because this seems like the most basic expected functionality for a personal finance tool... sorry but I have been disappointed by this shortcoming.

I guess I am just seeking guidance and to see if I am misunderstanding something. Thank you.

**Comment by JC5** [comment, marked as answer]
> So I should just make it a transfer in Firefly, right? 

Yes.

> Wouldn't it make more sense to have the ability for a transfer to link to both the withdrawal transaction, and the deposit transaction?

It's on my list to add a button for this, but since you're using CSV files: if you can map both ends of the transaction to the correct account, this will be done automatically.

> What happens if I import an updated csv, will I need to go back and delete that transaction again, to avoid doubled amounts?

No, it will not be imported again.

> I feel weird calling it an issue because this seems like the most basic expected functionality for a personal finance tool... sorry but I have been disappointed by this shortcoming.

Magically deleting or changing transactions has its drawbacks too, unfortunately.

**Reply by zzzachzzz** [reply]
>if you can map both ends of the transaction to the correct account, this will be done automatically

I did not realize this, that's good to know. Does this mean that the two transfers will be merged into one? When I tried converting both the withdrawal transaction and the deposit transaction to transfers, they remained as two transfers, which altered the account balance to be incorrect.

What are the criteria for this link to occur automatically? For one of my transfers from a bank checking account to a credit card as a payment, the dates don't quite line up. My credit card states `Transaction Date: 1/28/2023` & `Post Date: 1/29/2023`, while my bank account states: `Posting Date: 1/30/2023` & `Effective Date: 1/30/2023`. These clearly don't match and perhaps I mapped them incorrectly as well.

As seen here:
![Screenshot from 2023-04-02 11-11-11](https://user-images.githubusercontent.com/14324200/229368135-ce71db22-e416-453c-8eaa-38058502b76a.png)

**Reply by JC5** [reply]
> Does this mean that the two transfers will be merged into one?

That depends entirely on your bank. Most banks treat transfers as two transactions, and as a result, Firefly III will do so.

**Reply by zzzachzzz** [reply]
Right, in this case I'm talking about two totally separate accounts, one with gets a withdrawal transaction, and another that gets a deposit transaction. So the recommended approach then is still to convert one of these to a transfer, and delete the other?

I have not yet tried this alternative called Actual Budget, but this is how they handle it, making use of payees:
https://actualbudget.github.io/docs/Accounts/transfers/

**Reply by JC5** [reply]
Yes guess which app they copied that from. Firefly III calls it "source accounts" and "destination accounts" and if both are an asset account, it's a transfer. At that point, Actual Budget works just like Firefly III, not the other way around.

The problem is that Firefly III CANNOT assume that "these two transactions must be a single transfer, despite the fact they have different meta data in various fields, I'll just delete some shit and hope the user doesn't mind".

**Reply by zzzachzzz** [reply]
Okay. Ultimately this is what I'm seeking clarification on:
> >Wouldn't it make more sense to have the ability for a transfer to link to both the withdrawal transaction, and the deposit transaction?

>It's on my list to add a button for this, but since you're using CSV files: if you can map both ends of the transaction to the correct account, this will be done automatically.

I might be misunderstanding this automatic behavior you're referring to. As I understand it there are two options for making Firefly aware of a transfer, that takes the form of a withdrawal in one asset account, and a deposit in another asset account:
1. Convert one of the two transactions to a transfer, and delete the other
2. ? - What is the automatic behavior you're referring to. Is Firefly able to automatically create a "transfer", or a "link" (paid for by), etc. and prevent a duplicate that throws off the balance? Or does this process still require manual intervention each time

Edit: When you say 'map', are you referring to this feature? This is not a feature I have tried yet: https://docs.firefly-iii.org/data-importer/how-to-use/map/

Also I'm guessing this would need to be configured at import time

**Reply by JC5** [reply]
That's the one, yes.

**Comment by JcMinarro** [comment]
> > So I should just make it a transfer in Firefly, right?
> 
> Yes.
> 
> > Wouldn't it make more sense to have the ability for a transfer to link to both the withdrawal transaction, and the deposit transaction?
> 
> It's on my list to add a button for this, but since you're using CSV files: if you can map both ends of the transaction to the correct account, this will be done automatically.
> 
> > What happens if I import an updated csv, will I need to go back and delete that transaction again, to avoid doubled amounts?
> 
> No, it will not be imported again.

On my side, it is not the case.
I am importing multiple accounts from the same bank using Nordigen. The bank accounts are from Bankinter Bank.
Firefly maps properly the origin IBAN on the withdrawal and the destination IBAN on the deposit, but descriptions differ, they are:
```
# Deposit
/TXT/H|TRANS/SOME_NAME

#Withdrawal
/TXT/D|TRANS /SOME_NAME
```

So I understand Firefly is not able to merge them. Ok.
The problem comes after I categorize them. I convert one of them into a transaction and then remove the other one.
It looks like removing the transaction is removed from the DB (Even if I DON'T go to my profile and purge data) and then the next time it is imported from Data Importer (I am using External-ID Identifier as Duplicate Detection), it reappears on Firefly, the opposite that [docs](https://docs.firefly-iii.org/how-to/data-importer/import/duplicates/#deleted-but-duplicate) said.

Is it a bug? Or I am configuring anything on the wrong way? :thinking:

**Reply by JC5** [reply]
> It looks like removing the transaction is removed from the DB 

I doubt it. It should be easy to find the transaction in the `transaction_groups` table.

**Reply by JcMinarro** [reply]
> > It looks like removing the transaction is removed from the DB 
> 
> I doubt it. It should be easy to find the transaction in the `transaction_groups` table. 

I can confirm transactions imported from nordigen using the external identifier are permanently removed, and if you run the importer again, it is added again.
If the transactions are imported using a CSV file without an unique identifier and use the hash of the data, it is "hidden" instead of removed, and new import of the same transactions are not added.
If I go to my perfil and click on "purge" button then I import this "hidden" transaction, it is imported again

**Reply by JC5** [reply]
There is no code in Firefly III that I know of, that permanently deletes records on the first try. So I'm curious how this could happen.

**Reply by JcMinarro** [reply]
It is the Nordigen config I am using, and with this config, if I remove a transaction by clicking on "Delete" option and then the red button with the text "Delete Permanently" (I am attaching some pictures), when I run importer with the same account, transactions come again to firefly.
```
{
    "version": 3,
    "source": "ff3-importer-1.4.2",
    "created_at": "2024-01-23T01:16:11+01:00",
    "date": "",
    "default_account": 65,
    "delimiter": "comma",
    "headers": false,
    "rules": true,
    "skip_form": false,
    "add_import_tag": true,
    "roles": [],
    "do_mapping": [],
    "mapping": [
        {
            "***REMOVED SENSITIVE VALUE***": 92,
            "***REMOVED SENSITIVE VALUE***": 67,
            "***REMOVED SENSITIVE VALUE***": 55,
            "***REMOVED SENSITIVE VALUE***": 93,
            "***REMOVED SENSITIVE VALUE***": 96
        }
    ],
    "duplicate_detection_method": "cell",
    "ignore_duplicate_lines": false,
    "unique_column_index": 0,
    "unique_column_type": "external-id",
    "flow": "nordigen",
    "content_type": "csv",
    "custom_tag": "Bankinter-Importer",
    "identifier": "0",
    "connection": "0",
    "ignore_spectre_categories": false,
    "grouped_transaction_handling": "",
    "use_entire_opposing_address": false,
    "map_all_data": true,
    "accounts": {
        "***REMOVED SENSITIVE VALUE***": 65,
        "***REMOVED SENSITIVE VALUE***": 99,
        "***REMOVED SENSITIVE VALUE***": 67,
        "***REMOVED SENSITIVE VALUE***": 98
    },
    "date_range": "partial",
    "date_range_number": 23,
    "date_range_unit": "d",
    "date_not_before": "2023-12-31",
    "date_not_after": "",
    "nordigen_country": "ES",
    "nordigen_bank": "BANKINTER_BKBKESMM",
    "nordigen_requisitions": {
        "***REMOVED SENSITIVE VALUE***": "***REMOVED SENSITIVE VALUE***"
    },
    "nordigen_max_days": "90",
    "conversion": false,
    "ignore_duplicate_transactions": false
}
```
![Screenshot from 2024-01-28 18-01-17](https://github.com/firefly-iii/firefly-iii/assets/4047514/b09a1645-1af5-413c-8d90-9433e125746b)
![Screenshot from 2024-01-28 18-02-28](https://github.com/firefly-iii/firefly-iii/assets/4047514/f03dc94d-8ed3-4c03-b317-ddf3869d442e)



If I do the same "delete process" with a CSV import and using `"duplicate_detection_method": "classic"` on the config.json, it works as expected, the transactions is not re-added after it is removed.
